src/coreComponents/constitutive/solid/AuxiliaryFunctionsSpectral.hpp

```diff
@@ -0,0 +1,460 @@
+/*
+   0;10;1c * ------------------------------------------------------------------------------------------------------------
+ * SPDX-License-Identifier: LGPL-2.1-only
+ *
+ * Copyright (c) 2018-2020 Lawrence Livermore National Security LLC
+ * Copyright (c) 2018-2020 The Board of Trustees of the Leland Stanford Junior University
+ * Copyright (c) 2018-2020 Total, S.A
+ * Copyright (c) 2019-     GEOSX Contributors
+ * All rights reserved
+ *
+ * See top level LICENSE, COPYRIGHT, CONTRIBUTORS, NOTICE, and ACKNOWLEDGEMENTS files for details.
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+
+/**
+ * @file AuxiliaryFunctionsSpectral.hpp
+ */
+
+#ifndef GEOSX_CONSTITUTIVE_SOLID_AUXFUNSPECTRAL_HPP_
+#define GEOSX_CONSTITUTIVE_SOLID_AUXFUNSPECTRAL_HPP_
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include "LvArray/src/output.hpp"
+#include "LvArray/src/tensorOps.hpp"
+namespace geosx
+{
+
+//get the positive part only of tensor T using spectral split
+GEOSX_HOST_DEVICE inline
+void PositivePartOfTensor( real64 (& eigs)[3], real64 (& eigvecs)[3][3], real64 (& positivePart)[6] )
+{
+  real64 positiveEigs[6]={};
+  for( int i=0; i < 3; i++ )
+  {
+    positiveEigs[i] = std::max( 0.0, eigs[i] );
+  }
+  LvArray::tensorOps::Rij_eq_AikSymBklAjl< 3 >( positivePart, eigvecs, positiveEigs );
+}
+
+GEOSX_HOST_DEVICE inline
+void NegativePartOfTensor( real64 (& eigs)[3], real64 (& eigvecs)[3][3], real64 (& negativePart)[6] )
+{
+  real64 negativeEigs[6]={};
+  for( int i=0; i < 3; i++ )
+  {
+    negativeEigs[i] = std::min( 0.0, eigs[i] );
+  }
+  LvArray::tensorOps::Rij_eq_AikSymBklAjl< 3 >( negativePart, eigvecs, negativeEigs );
+}
+
+GEOSX_HOST_DEVICE inline
+real64 doubleContraction( real64 (& A)[6], real64 (& B)[6] )
+{
+  real64 ans = 0;
+  for( int i=0; i < 6; i++ )
+  {
+    if( i < 3 )
+    {
+      ans = ans + A[i]*B[i];
+    }
+    else
+    {
+      ans = ans + 2*A[i]*B[i];
+    }
+  }
+  return ans;
+}
+
+GEOSX_HOST_DEVICE inline
+void recoverStrainFromStress( arraySlice1d< real64 > const stress, real64 (& strain)[6], real64 const K, real64 const mu )
+{
+  real64 E = 9*K*mu / (3*K + mu);
+  real64 nu = (3*K - 2*mu) / (6*K + 2*mu);
+  strain[0] = (stress[0] - nu*(stress[1] + stress[2]))/E;
+  strain[1] = (stress[1] - nu*(stress[0] + stress[2]))/E;
+  strain[2] = (stress[2] - nu*(stress[0] + stress[1]))/E;
+  strain[3] = (1 + nu)*stress[3]/E;
+  strain[4] = (1 + nu)*stress[4]/E;
+  strain[5] = (1 + nu)*stress[5]/E;
+}
+
+GEOSX_HOST_DEVICE inline
+int voigt( int voigtIndex, int pairIndex )
```

Comment:
why do we have this? This is a very inefficient operation.

Reply:
I need to convert from matrix notation to voigt notation in some functions. That was the simplest way I could think of.